test/test_traceradvectiondiffusion.jl

```diff
@@ -145,6 +145,86 @@ function test_timedependentvel(stepper, dt, tfinal, dev::Device=CPU(); uvel = 0.
   return isapprox(cfinal, vs.c, rtol = gr.nx*gr.ny*nsteps*1e-12)
 end
 
+
+"""
+    test_constvel3D(; kwargs...)
+
+Advect a gaussian concentration `c0(x, y, t)` with a constant velocity flow
+`u(x, y, z) = uvel`, `v(x, y, z) = vvel` and `w(x, y, z) = wvel` and compares the final state with
+`cfinal = c0(x - uvel * tfinal, y - vvel * tfinal, z - wvel * tfinal)`.
+"""
+function test_constvel3D(stepper, dt, nsteps, dev::Device=CPU())
+
+  nx, Lx = 128, 2π
+  uvel, vvel, wvel = 0.2, 0.1, 0.15
+  u(x, y, z) = uvel
+  v(x, y, z) = vvel
+  w(x, y, z) = wvel
+  advecting_flow = ThreeDAdvectingFlow(; u, v, w)
+
+  prob = TracerAdvectionDiffusion.Problem(dev, advecting_flow; nx, Lx, κ=0.0, dt, stepper)
+  sol, cl, vs, pr, gr = prob.sol, prob.clock, prob.vars, prob.params, prob.grid
+
+  x, y, z = gridpoints(gr)
+
+  σ = 0.1
+  c0ampl = 0.1
+  c0func(x, y, z) = c0ampl * exp(-(x^2 + y^2 + z^2) / 2σ^2)
+
+  c0 = c0func.(x, y, z)
+  tfinal = nsteps * dt
+  cfinal = @. c0func(x - uvel * tfinal, y - vvel * tfinal, z - wvel * tfinal)
+
+  TracerAdvectionDiffusion.set_c!(prob, c0)
+
+  stepforward!(prob, nsteps)
+  TracerAdvectionDiffusion.updatevars!(prob)
+
+  return isapprox(cfinal, vs.c, rtol = gr.nx*gr.ny*gr.nz*nsteps*1e-12)
+end
+
+
+"""
+    test_timedependenttvel3D(; kwargs...)
+
+Advect a gaussian concentration `c0(x, y, z, t)` with a time-varying velocity flow
+`u(x, y, z, t) = uvel`, `v(x, y, z, t) = vvel * sign(-t + tfinal/2)` and `w(x, y, z, t) = wvel` and
+compares the final state with `cfinal = c0(x - uvel * tfinal, y, z - wvel * tfinal)`.
+"""
+function test_timedependentvel3D(stepper, dt, tfinal, dev::Device=CPU(); uvel = 0.5, αv = 0.5, wvel = 0.5)
+  
+  nx, Lx = 128, 2π
+  nsteps = round(Int, tfinal/dt)
+  
+  if !isapprox(tfinal, nsteps*dt, rtol=rtol_traceradvectiondiffusion)
+    error("tfinal is not multiple of dt")
+  end
+  
+  u(x, y, z, t) = uvel
+  v(x, y, z, t) = αv * t + αv * dt/2
```

Comment:
why do we need `+ αv * dt/2` here?

(It's in other tests... but I don't remember why it was needed there in the first place...)

Reply:
Does it make it effectively leapfrog scheme?